### PR TITLE
Add support for parsing log entries w/ 'protoPayload' from resources

### PR DESCRIPTION
**Pull request by tseaver**
Uses #1574 as a base.

Note that we can't feasibly allow writing such entries until we have an answer for #1577.

### CODE REVIEW

**Comment by tseaver** [comment]
@dhermes 8d5ae2b is the commit which disables the `too-many-lines`/`too-many-statements` limits altogether for test code.  That feels cleaner to me than either disabling the limit file-by-file, or repeatedly nudging it upwards.

**Comment by dhermes** [comment]
@tseaver `too-many-lines` / `too-many-statements` should not be disabled. They provide valuable feedback when we've maybe gone a little too far. e.g. when the `row.py` cartel should've been broken up (#1602)

File: gcloud/logging/entries.py [hunk 1]
```diff
@@ -85,9 +85,18 @@ class TextEntry(_BaseEntry):
...
     _PAYLOAD_KEY = 'jsonPayload'
+
+
+class ProtobufEntry(_BaseEntry):
+    """Entry created with ``protoPayload``.
+
+    See:
+    https://cloud.google.com/logging/docs/api/ref_v2beta1/rest/v2beta1/LogEntry
+    """
+    _PAYLOAD_KEY = 'protoPayload'
```
Comment (dhermes) [inline gcloud/logging/entries.py]: So this doesn't try to parse the payload anywhere?

File: gcloud/logging/test_client.py [hunk 2]
```diff
@@ -107,7 +108,10 @@ def test_list_entries_explicit(self):
         NOW = datetime.utcnow().replace(tzinfo=UTC)
         TIMESTAMP = _datetime_to_rfc3339_w_nanos(NOW)
         IID1 = 'IID1'
+        IID2 = 'IID2'
         PAYLOAD = {'message': 'MESSAGE', 'weather': 'partly cloudy'}
+        PROTO_PAYLOAD = PAYLOAD.copy()
+        PROTO_PAYLOAD['@type'] = 'type.googleapis.com/testing.example'
```
Comment (dhermes) [inline gcloud/logging/test_client.py]: You found docs or an example somewhere? Link?

File: gcloud/logging/test_entries.py [hunk 3]
```diff
@@ -221,6 +221,114 @@ def test_from_api_repr_w_loggers_w_logger_match(self):
         self.assertTrue(entry.logger is LOGGER)
 
 
+class TestProtobufEntry(unittest2.TestCase):
```
Comment (dhermes) [inline gcloud/logging/test_entries.py]: Do we need a test this big for a "do-nothing" subclass?

**Comment by dhermes** [comment]
I take it this hasn't been rebased yet?

File: gcloud/logging/entries.py (hunk 1 above)
Comment (tseaver) [inline gcloud/logging/entries.py]: It is already parsed:  although the *name* says `protPayload`, it actaully comes through as JSON.

File: gcloud/logging/test_client.py (hunk 2 above)
Comment (tseaver) [inline gcloud/logging/test_client.py]: Only dummied up versions, e.g. https://cloud.google.com/logging/docs/api/ref_v2beta1/rest/v2beta1/LogEntry#FIELDS.proto_payload

Comment (dhermes) [inline gcloud/logging/test_client.py]: It'd be nice to see one in the wild. How does the payload get set?

**Comment by tseaver** [comment]
\> too-many-lines / too-many-statements

For testing code, I disagree:  "it takes as much as it takes" to cover the MUT, and the arbitrary restriction on testcase method length tends to lead to "getting fancy" in testing code.

File: gcloud/logging/test_entries.py (hunk 3 above)
Comment (tseaver) [inline gcloud/logging/test_entries.py]: It follows the other two:  we aren't testing the base class explicitly.

**Comment by tseaver** [comment]
> I take it this hasn't been rebased yet?

It is current with the tip of the `logging-api` branch.

**Comment by dhermes** [comment]
RE: Rebased, the "Add 'Logger.list_entries'." commit must not have merged cleanly then. Maybe you want to cherry-pick the last two commits?

**Comment by tseaver** [comment]
Github seems confused: e4b4b0d is already [present on the `logging-api` branch](https://github.com/GoogleCloudPlatform/gcloud-python/commits/logging-api),  When I try to rebase against `logging-api` in my local checkout, that commit does not appear anywhere.

**Comment by dhermes** [comment]
Try a bit more to unconfuse it? Maybe re-fetch the remote locally?

**Comment by tseaver** [comment]
How?  The histories look identical to me:

```
$ git pull upstream logging-api 
From github.com:GoogleCloudPlatform/gcloud-python
 * branch            logging-api -> FETCH_HEAD
Already up-to-date.

$ git log -3 upstream/logging-api
commit b8151b1e0f7dbf7b9b5a5beff65ca49042798a99
Merge: 68baa1b e4b4b0d
Author: Tres Seaver <tseaver@palladion.com>
Date:   Fri Mar 11 12:44:00 2016 -0500

    Merge pull request #1574 from tseaver/logging-logger_list_entries
    
    Add 'Logger.list_entries'.

commit e4b4b0d2d1a6bfb9348ff2f073a9577cdf03c378
Author: Tres Seaver <tseaver@palladion.com>
Date:   Wed Mar 9 16:17:14 2016 -0500

    Add 'Logger.list_entries'.

commit 68baa1b05ac0b15e608397756790bb5f06d14846
Merge: e898bca d7a2e72
Author: Tres Seaver <tseaver@palladion.com>
Date:   Fri Mar 11 11:55:02 2016 -0500

    Merge pull request #1569 from tseaver/logging-client_list_entries
    
    Add 'Client.list_entries'.

$ git log -3 logging-api
commit b8151b1e0f7dbf7b9b5a5beff65ca49042798a99
Merge: 68baa1b e4b4b0d
Author: Tres Seaver <tseaver@palladion.com>
Date:   Fri Mar 11 12:44:00 2016 -0500

    Merge pull request #1574 from tseaver/logging-logger_list_entries
    
    Add 'Logger.list_entries'.

commit e4b4b0d2d1a6bfb9348ff2f073a9577cdf03c378
Author: Tres Seaver <tseaver@palladion.com>
Date:   Wed Mar 9 16:17:14 2016 -0500

    Add 'Logger.list_entries'.

commit 68baa1b05ac0b15e608397756790bb5f06d14846
Merge: e898bca d7a2e72
Author: Tres Seaver <tseaver@palladion.com>
Date:   Fri Mar 11 11:55:02 2016 -0500

    Merge pull request #1569 from tseaver/logging-client_list_entries
    
    Add 'Client.list_entries'.

$ git log -5 logging-parse_protobuf_entries 
commit 124ccf08c929794f2cd4ba5ba9fef3d9fdbf42d3
Author: Tres Seaver <tseaver@palladion.com>
Date:   Mon Mar 7 11:37:16 2016 -0500

    Add support for parsing log entries w/ 'protoPayload' from resources.
    
    We can't feasibly allow writing such entries until we have an answer for

commit 465699c255e39132cc075e114f44807150d9cce2
Author: Tres Seaver <tseaver@palladion.com>
Date:   Mon Mar 7 11:35:57 2016 -0500

    Disable 'too-many-statements'/'too-many-lines' for test code.

commit b8151b1e0f7dbf7b9b5a5beff65ca49042798a99
Merge: 68baa1b e4b4b0d
Author: Tres Seaver <tseaver@palladion.com>
Date:   Fri Mar 11 12:44:00 2016 -0500

    Merge pull request #1574 from tseaver/logging-logger_list_entries
    
    Add 'Logger.list_entries'.

commit e4b4b0d2d1a6bfb9348ff2f073a9577cdf03c378
Author: Tres Seaver <tseaver@palladion.com>
Date:   Wed Mar 9 16:17:14 2016 -0500

    Add 'Logger.list_entries'.

commit 68baa1b05ac0b15e608397756790bb5f06d14846
Merge: e898bca d7a2e72
Author: Tres Seaver <tseaver@palladion.com>
Date:   Fri Mar 11 11:55:02 2016 -0500

    Merge pull request #1569 from tseaver/logging-client_list_entries
    
    Add 'Client.list_entries'.
```

**Comment by dhermes** [comment]
Looks like you've sorted it out. Only two commits now in this PR.

**Comment by tseaver** [comment]
I just tried a new `push -f` from the `loggin-parse_protobuf_entries` branch, which seems to have cleared the stray commit.

**Comment by dhermes** [comment]
> For testing code, I disagree: "it takes as much as it takes" to cover the MUT, and the arbitrary restriction on testcase method length tends to lead to "getting fancy" in testing code.

It's not hard to argue that code like the [lint-disabled `test_upload_from_file_w_explicit_client_resumable`][1] is hard to follow. These restrictions give us a signal which is valuable to us when we get too deep in our own code and lose context.

[1]: https://github.com/GoogleCloudPlatform/gcloud-python/blob/97eb7f5cac557f17ecd5311c2f374d7eaa56f8ed/gcloud/bigquery/test_table.py#L1405-L1489

**Comment by tseaver** [comment]
@dhermes 7c52009 drops suppressing the project-wide 'too-many-statements' check for test code, and instead disables it locally for the one newly-added test:  [looking at it](https://github.com/GoogleCloudPlatform/gcloud-python/pull/1578/files#diff-03e0b74a990a1b9edbca70608de3abb9L106), I don't see any reason to break it up.  PTAL

File: gcloud/logging/test_entries.py (hunk 3 above)
Comment (dhermes) [inline gcloud/logging/test_entries.py]: Why not test the base-class explicitly then?

**Comment by dhermes** [comment]
Only real pending issue is my comment about seeing an example of a proto payload from a real API request. Have you been able to produce one?

**Comment by tseaver** [comment]
From the ["Try It" form](https://cloud.google.com/logging/docs/api/ref_v2beta1/rest/v2beta1/entries/list):

```
POST https://logging.googleapis.com/v2beta1/entries:list?key={YOUR_API_KEY}

{
 "projectIds": [
  "citric-celerity-697"
 ]
}
```

returns:

```
200 OK

- Hide headers -

Cache-Control:  private
Content-Encoding:  gzip
Content-Length:  5768
Content-Type:  application/json; charset=UTF-8
Date:  Tue, 22 Mar 2016 01:52:27 GMT
Server:  ESF
Vary:  Origin, X-Origin, Referer

{
 "entries": [
  {
   "textPayload": "Logging from gcloud-python!",
   "insertId": "2016-03-03|11:28:12.764719-08|10.95.84.10|1527583067",
   "resource": {
    "type": "global"
   },
   "timestamp": "2016-03-03T19:28:12.750563482Z",
   "logName": "projects/citric-celerity-697/logs/hacking-around"
  },
  {
   "insertId": "2016-03-03|11:48:07.740612-08|10.95.212.81|-486515530",
   "jsonPayload": {
    "weather": "cloudy",
    "message": "Now with arbitrary key-values!"
   },
   "resource": {
    "type": "global"
   },
   "timestamp": "2016-03-03T19:48:07.728837588Z",
   "logName": "projects/citric-celerity-697/logs/hacking-around"
  },
  {
   "protoPayload": {
    "@type": "type.googleapis.com/google.cloud.audit.AuditLog",
    "status": {
    },
    "authenticationInfo": {
    },
    "requestMetadata": {
    },
    "serviceName": "bigquery.googleapis.com",
    "methodName": "datasetservice.list",
    "authorizationInfo": [
     {
      "resource": "projects/citric-celerity-697",
      "permission": "bigquery.datasets.list",
      "granted": true
     },
     {
      "resource": "projects/citric-celerity-697/datasets/test_listing_datasets",
      "permission": "bigquery.datasets.get",
      "granted": true
     }
    ],
    "resourceName": "projects/citric-celerity-697/datasets",
    "serviceData": {
     "@type": "type.googleapis.com/google.cloud.bigquery.logging.v1.AuditData",
     "datasetListRequest": {
     }
    }
   },
   "insertId": "52D3EC4A89C80.A5FD1C7.7C0038C9",
   "resource": {
    "type": "bigquery_resource"
   },
   "timestamp": "2016-03-04T20:14:54.160Z",
   "severity": "INFO",
   "logName": "projects/citric-celerity-697/logs/cloudaudit.googleapis.com%2Fdata_access"
  },
...
 ]
}
```

**Comment by dhermes** [comment]
Really helpful! Thanks

**Comment by dhermes** [comment]
LGTM (though I still think reducing the test boilerplate for the do-nothing subclasses would be a net positive)

**Comment by tseaver** [comment]
@dhermes I changed `gcloud.logging.test_entries` to test `_BaseEntry` directly, dropping tests for the method-less derived classes.  I will merge when Travis passes.

**Comment by dhermes** [comment]
SGTM thanks
